tracing-test-utils/src/main/java/com/palantir/tracing/SpanRenderer.java

```diff
@@ -0,0 +1,382 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.graph.EndpointPair;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import com.google.common.hash.Hashing;
+import com.palantir.tracing.api.Span;
+import com.palantir.tracing.api.SpanObserver;
+import com.palantir.tracing.api.SpanType;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Clock;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings({"PreferSafeLoggableExceptions", "Slf4jLogsafeArgs"}) // test-lib, no need for SafeArgs
+final class SpanRenderer implements SpanObserver {
+
+    private static final Logger log = LoggerFactory.getLogger(SpanRenderer.class);
+    private final Collection<Span> allSpans = new ArrayBlockingQueue<>(1000);
+
+    @Override
+    public void consume(Span span) {
+        allSpans.add(span);
+    }
+
+    @SuppressWarnings({"JavaTimeDefaultTimeZone", "BanSystemOut"}) // I actually want the system default time zone!
+    void output(String displayName, Path path) {
+        TimeBounds bounds = bounds(allSpans);
+
+        Map<String, List<Span>> spansByTraceId = allSpans.stream()
+                .collect(Collectors.groupingBy(Span::getTraceId));
+
+        Map<String, AnalyzedSpans> analyzedByTraceId = Maps.transformValues(spansByTraceId, SpanRenderer::analyze);
+
+        HtmlFormatter formatter = new HtmlFormatter(bounds);
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("<h1>");
+        sb.append(displayName);
+        sb.append("</h1>");
+        sb.append("<p>");
+        sb.append(DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss")
+                .format(LocalDateTime.now(Clock.systemDefaultZone())));
+        sb.append("</p>");
+        analyzedByTraceId.entrySet()
+                .stream()
+                .sorted(Comparator.comparingLong(e1 -> e1.getValue().bounds().startMicros()))
+                .forEachOrdered(entry -> {
+                    AnalyzedSpans analysis = entry.getValue();
+                    formatter.renderAllSpansForOneTraceId(entry.getKey(), analysis, sb);
+                });
+
+        formatter.rawSpanJson(allSpans, sb);
+
+        try {
+            Files.write(
+                    path,
+                    sb.toString().getBytes(StandardCharsets.UTF_8));
```

Comment:
`sb` contains the full formatted output on heap
`sb.toString()` creates a second copy
`getBytes` create a third copy